pkg/services/hcloud/loadbalancer/loadbalancer.go

```diff
@@ -138,163 +133,152 @@ func (s *Service) reconcileNetworkAttachement(ctx context.Context, lb *hcloud.Lo
 			if _, err := s.scope.HCloudClient.DeleteServiceFromLoadBalancer(ctx, lb, listenPort); err != nil {
+				// return immediately on rate limit
 				hcloudutil.HandleRateLimitExceeded(s.scope.HetznerCluster, err, "DeleteServiceFromLoadBalancer")
-				multierr = append(multierr, fmt.Errorf("error deleting service from load balancer: %s", err))
+				multierr = append(multierr, fmt.Errorf("failed to delete service from load balancer: %w", err))
+				if hcloud.IsError(err, hcloud.ErrorCodeRateLimitExceeded) {
+					return kerrors.NewAggregate(multierr)
+				}
 			}
 		}
 	}
 
-	// Create services which are in specs and not yet in API
+	// create services which are in specs and not yet in API
 	for i, listenPort := range toCreate {
 		proxyProtocol := false
-		destinationPort := specServiceListenPortsMap[listenPort].DestinationPort
+		destinationPort := wantServiceListenPortsMap[listenPort].DestinationPort
 		serviceOpts := hcloud.LoadBalancerAddServiceOpts{
-			Protocol:        hcloud.LoadBalancerServiceProtocol(specServiceListenPortsMap[listenPort].Protocol),
+			Protocol:        hcloud.LoadBalancerServiceProtocol(wantServiceListenPortsMap[listenPort].Protocol),
 			ListenPort:      &toCreate[i],
 			DestinationPort: &destinationPort,
 			Proxyprotocol:   &proxyProtocol,
 		}
 		if _, err := s.scope.HCloudClient.AddServiceToLoadBalancer(ctx, lb, serviceOpts); err != nil {
+			// return immediately on rate limit
 			hcloudutil.HandleRateLimitExceeded(s.scope.HetznerCluster, err, "AddServiceToLoadBalancer")
-			multierr = append(multierr, fmt.Errorf("error adding service to load balancer: %s", err))
+			multierr = append(multierr, fmt.Errorf("failed to add service to load balancer: %w", err))
+			if hcloud.IsError(err, hcloud.ErrorCodeRateLimitExceeded) {
+				return kerrors.NewAggregate(multierr)
+			}
 		}
 	}
-
 	return kerrors.NewAggregate(multierr)
 }
 
-func (s *Service) createLoadBalancer(ctx context.Context) (*hcloud.LoadBalancer, error) {
-	log := ctrl.LoggerFrom(ctx)
-
-	log.Info("Create a new loadbalancer", "algorithm type", s.scope.HetznerCluster.Spec.ControlPlaneLoadBalancer.Algorithm)
+func max(x, y int) int {
+	if x > y {
+		return x
+	}
+	return y
+}
 
-	res, err := s.scope.HCloudClient.CreateLoadBalancer(ctx, buildLoadBalancerCreateOpts(s.scope.HetznerCluster))
+func (s *Service) createLoadBalancer(ctx context.Context) (*hcloud.LoadBalancer, error) {
+	opts := createOptsFromSpec(s.scope.HetznerCluster)
+	res, err := s.scope.HCloudClient.CreateLoadBalancer(ctx, opts)
 	if err != nil {
 		hcloudutil.HandleRateLimitExceeded(s.scope.HetznerCluster, err, "CreateLoadBalancer")
-		record.Warnf(
-			s.scope.HetznerCluster,
-			"FailedCreateLoadBalancer",
-			"Failed to create load balancer: %s",
-			err)
-		return nil, fmt.Errorf("error creating load balancer: %w", err)
+		reterr := fmt.Errorf("failed to create load balancer: %w", err)
+		record.Warnf(s.scope.HetznerCluster, "FailedCreateLoadBalancer", reterr.Error())
+		return nil, reterr
```

Comment:
What do you think about using this?

`err := fmt.Errorf("failed to create load balancer: %w", err)` instead of creating a `reterr`?

There are several places with this pattern.

Reply:
I did this because err already exists there. I didn't want to "secretely overwrite". But I am open to do that if you think it is better

Reply:
reading `err` feels a bit natural for me. If the value is `reterr` then it looks a bit strange. I personally think overwriting `err` makes sense in this case. It is not a different err. It is the same err, but wrapped with an additional message.

Reply:
okay, I will update it!